In [1]:
#pip install yfinance
import pandas  as pd #Data manipulation
import numpy as np #Data manipulation

import matplotlib.pyplot as plt # plot graphics
plt.style.use("seaborn")

import yfinance as yf # Yahoo finance
from datetime import datetime

In [2]:
# my portfolio
# "BIMBOA.MX", "HERDEZ.MX", "BOLSAA.MX",
# yf.download(tickers = symbol, start = start, end = end)
symbol = ["GENIUS21.MX", "IVVPESOISHRS.MX", "WALMEX.MX", "AAPL", "AMZN", "FDX","NVDA", "XOM"]
start = "2023-01-02" 
# manyways to say now
end = datetime.now().strftime("%Y-%m-%d")

portfolio_df = yf.download(tickers=symbol, start = start, end = end)
portfolio_df.head()
portfolio_df.info()

[*********************100%%**********************]  8 of 8 completed


<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 399 entries, 2023-01-02 to 2024-07-17
Data columns (total 48 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   (Adj Close, AAPL)             386 non-null    float64
 1   (Adj Close, AMZN)             386 non-null    float64
 2   (Adj Close, FDX)              386 non-null    float64
 3   (Adj Close, GENIUS21.MX)      384 non-null    float64
 4   (Adj Close, IVVPESOISHRS.MX)  381 non-null    float64
 5   (Adj Close, NVDA)             386 non-null    float64
 6   (Adj Close, WALMEX.MX)        388 non-null    float64
 7   (Adj Close, XOM)              386 non-null    float64
 8   (Close, AAPL)                 386 non-null    float64
 9   (Close, AMZN)                 386 non-null    float64
 10  (Close, FDX)                  386 non-null    float64
 11  (Close, GENIUS21.MX)          384 non-null    float64
 12  (Close, IVVPESOISHRS.MX)      381 non-null   

In [7]:
# get AAPL stock info
#portfolio_close_df = portfolio_df['Close']
#aapl_df = portfolio_close_df['AAPL']
aapl_df = portfolio_df['Close']['AAPL']
aapl_df

Date
2023-01-02           NaN
2023-01-03    125.070000
2023-01-04    126.360001
2023-01-05    125.019997
2023-01-06    129.619995
                 ...    
2024-07-11    227.570007
2024-07-12    230.539993
2024-07-15    234.399994
2024-07-16    234.820007
2024-07-17    228.880005
Name: AAPL, Length: 399, dtype: float64

In [4]:
# rest of stocks
amzn_df   = portfolio_df['Close']['AMZN']
fdx_df    = portfolio_df['Close']['FDX']
nvda_df   = portfolio_df['Close']['NVDA']
xom_df    = portfolio_df['Close']['XOM']
gen21_df  = portfolio_df['Close']['GENIUS21.MX']
ivv_df    = portfolio_df['Close']['IVVPESOISHRS.MX']
walmex_df = portfolio_df['Close']['WALMEX.MX']

In [8]:
# defining two data sets, daily and monthly
# https://stackoverflow.com/questions/51604967/return-last-date-and-value-each-month-in-pandas
aapl_df_m = aapl_df.groupby([aapl_df.index.year,aapl_df.index.month]).tail(1)
aapl_df_m

Date
2023-01-31    144.289993
2023-02-28    147.410004
2023-03-31    164.899994
2023-04-28    169.679993
2023-05-31    177.250000
2023-06-30    193.970001
2023-07-31    196.449997
2023-08-31    187.869995
2023-09-29    171.210007
2023-10-31    170.770004
2023-11-30    189.949997
2023-12-29    192.529999
2024-01-31    184.399994
2024-02-29    180.750000
2024-03-28    171.479996
2024-04-30    170.330002
2024-05-31    192.250000
2024-06-28    210.619995
2024-07-17    228.880005
Name: AAPL, dtype: float64

In [9]:
# rest of stocks
amzn_df_m   = amzn_df.groupby([amzn_df.index.year, amzn_df.index.month]).tail(1)
fdx_df_m    = fdx_df.groupby([fdx_df.index.year, fdx_df.index.month]).tail(1)
nvda_df_m   = nvda_df.groupby([nvda_df.index.year, nvda_df.index.month]).tail(1)
xom_df_m    = xom_df.groupby([xom_df.index.year, xom_df.index.month]).tail(1)
gen21_df_m  = gen21_df.groupby([gen21_df.index.year, gen21_df.index.month]).tail(1)
ivv_df_m    = ivv_df.groupby([ivv_df.index.year, ivv_df.index.month]).tail(1)
walmex_df_m = walmex_df.groupby([walmex_df.index.year, walmex_df.index.month]).tail(1)

In [10]:
# interactive plots cufflinks 
# import it or install it
try:
    import cufflinks as cf
except:
    %pip install cufflinks
    import cufflinks as cf

In [11]:
cf.set_config_file(offline = True)
aapl_df.iplot()

In [12]:
# creating US df with close values
us_portfolio_df = pd.DataFrame()
us_portfolio_df['AAPL'] = aapl_df.dropna()
us_portfolio_df['AMZN'] = amzn_df.dropna()
us_portfolio_df['FDX']  = fdx_df.dropna()
us_portfolio_df['NVDA'] = nvda_df.dropna()
us_portfolio_df['XOM']  = xom_df.dropna()

# https://numpy.org/doc/stable/reference/generated/numpy.log.html
#us_portfolio_df["AAPL_Returns"] = np.log(aapl_df.div(aapl_df.shift(1)))
us_portfolio_df["AAPL_Returns"] = aapl_df.pct_change()
#us_portfolio_df["AMZN_Returns"] = np.log(amzn_df.div(amzn_df.shift(1)))
us_portfolio_df["AMZN_Returns"] = amzn_df.pct_change()
#us_portfolio_df["FDX_Returns"] = np.log(fdx_df.div(fdx_df.shift(1)))
us_portfolio_df["FDX_Returns"] = fdx_df.pct_change()
#us_portfolio_df["NVDA_Returns"] = np.log(nvda_df.div(nvda_df.shift(1)))
us_portfolio_df["NVDA_Returns"] = nvda_df.pct_change()
#us_portfolio_df["XOM_Returns"] = np.log(xom_df.div(xom_df.shift(1)))
us_portfolio_df["XOM_Returns"] = xom_df.pct_change()

us_portfolio_df

AAPL        AMZN         FDX        NVDA         XOM  \
Date                                                                     
2023-01-03  125.070000   85.820000  177.270004   14.315000  106.510002   
2023-01-04  126.360001   85.139999  181.449997   14.749000  106.820000   
2023-01-05  125.019997   83.120003  181.279999   14.265000  109.209999   
2023-01-06  129.619995   86.080002  185.770004   14.859000  110.529999   
2023-01-09  130.149994   87.360001  189.740005   15.628000  108.470001   
...                ...         ...         ...         ...         ...   
2024-07-11  227.570007  195.050003  299.540009  127.400002  113.250000   
2024-07-12  230.539993  194.490005  299.019989  129.240005  113.269997   
2024-07-15  234.399994  192.720001  305.019989  128.440002  115.209999   
2024-07-16  234.820007  193.020004  313.519989  126.360001  116.040001   
2024-07-17  228.880005  187.929993  310.429993  117.989998  117.639999   

            AAPL_Returns  AMZN_Returns  FDX_Returns  NVDA_Returns  XOM_Returns  
Date                                                                            
2023-01-03           NaN           NaN          NaN           NaN          NaN  
2023-01-04      0.010314     -0.007924     0.023580      0.030318     0.002911  
2023-01-05     -0.010605     -0.023726    -0.000937     -0.032816     0.022374  
2023-01-06      0.036794      0.035611     0.024768      0.041640     0.012087  
2023-01-09      0.004089      0.014870     0.021371      0.051753    -0.018637  
...                  ...           ...          ...           ...          ...  
2024-07-11     -0.023221     -0.023725     0.021380     -0.055667     0.011884  
2024-07-12      0.013051     -0.002871    -0.001736      0.014443     0.000177  
2024-07-15      0.016743     -0.009101     0.020066     -0.006190     0.017127  
2024-07-16      0.001792      0.001557     0.027867     -0.016194     0.007204  
2024-07-17     -0.025296     -0.026370    -0.009856     -0.066239     0.013788  

[386 rows x 10 columns]

In [13]:
# create US dayly df difference with values
us_portfolio_diff_df = pd.DataFrame()
ticks_us = ['AAPL', 'AMZN', 'FDX', 'NVDA', 'XOM']

for t in ticks_us:
    # transform into differences dt
    tmp_l = []
    j = 0
    for i in us_portfolio_df[t].index:
        #print(t, i, j)
        if j > 0:
            # look for consecutive cells
            #print(us_portfolio_df[t].iloc[j-1] - us_portfolio_df[t].iloc[j])
            tmp_l.append(us_portfolio_df[t].iloc[j] - us_portfolio_df[t].iloc[j-1])
        j = j + 1
    us_portfolio_diff_df = pd.concat([us_portfolio_diff_df, pd.DataFrame(tmp_l, columns=[t])], axis=1)

#tmp_portfolio_diff_df = pd.DataFrame(tmp_l, columns=[t])
#us_portfolio_diff_df = pd.concat([us_portfolio_diff_df, tmp_portfolio_diff_df], ignore_index=True, sort=False)

us_portfolio_diff_df

AAPL      AMZN       FDX      NVDA       XOM
0    1.290001 -0.680000  4.179993  0.434000  0.309998
1   -1.340004 -2.019997 -0.169998 -0.483999  2.389999
2    4.599998  2.959999  4.490005  0.594000  1.320000
3    0.529999  1.279999  3.970001  0.769000 -2.059998
4    0.580002  2.510002 -1.000000  0.281000  1.619995
..        ...       ...       ...       ...       ...
380 -5.409988 -4.739990  6.270020 -7.510002  1.330002
381  2.969986 -0.559998 -0.520020  1.840004  0.019997
382  3.860001 -1.770004  6.000000 -0.800003  1.940002
383  0.420013  0.300003  8.500000 -2.080002  0.830002
384 -5.940002 -5.090012 -3.089996 -8.370003  1.599998

[385 rows x 5 columns]

In [14]:
# plot us portfolio
cf.set_config_file(offline = True)
us_portfolio_diff_df.iplot()

In [19]:
# see if data is normal
# https://github.com/santosjorge/cufflinks/blob/master/Cufflinks%20Tutorial%20-%20Plotly.ipynb
ticks_us = ['AAPL', 'AMZN', 'FDX', 'NVDA', 'XOM']
#us_portfolio_diff_df.sort_values(ticks_us, ascending=[True, True, True, True, True]).iplot(kind='bar')
us_portfolio_diff_df[ticks_us].iplot(kind='bar')

In [20]:
# plot us portfolio
cf.set_config_file(offline = True)
us_portfolio_df[ticks_us].iplot()

In [21]:
# create US monthly df difference with values
us_portfolio_m_df = pd.DataFrame()
us_portfolio_m_diff_df = pd.DataFrame()
ticks_us = ['AAPL', 'AMZN', 'FDX', 'NVDA', 'XOM']

us_portfolio_m_df['AAPL'] = aapl_df_m.dropna()
us_portfolio_m_df['AMZN'] = amzn_df_m.dropna()
us_portfolio_m_df['FDX']  = fdx_df_m.dropna()
us_portfolio_m_df['NVDA'] = nvda_df_m.dropna()
us_portfolio_m_df['XOM']  = xom_df_m.dropna()

for t in ticks_us:
    # transform into differences dt
    tmp_l = []
    j = 0
    for i in us_portfolio_m_df[t].index:
        #print(t, i, j)
        if j > 0:
            # look for consecutive cells
            #print(us_portfolio_df[t].iloc[j-1] - us_portfolio_df[t].iloc[j])
            tmp_l.append(us_portfolio_m_df[t].iloc[j] - us_portfolio_m_df[t].iloc[j-1])
        j = j + 1
    us_portfolio_m_diff_df = pd.concat([us_portfolio_m_diff_df, pd.DataFrame(tmp_l, columns=[t])], axis=1)

#tmp_portfolio_diff_df = pd.DataFrame(tmp_l, columns=[t])
#us_portfolio_diff_df = pd.concat([us_portfolio_diff_df, tmp_portfolio_diff_df], ignore_index=True, sort=False)

us_portfolio_m_diff_df

AAPL       AMZN        FDX       NVDA        XOM
0    3.120010  -8.899994   9.360001   3.678999  -6.099998
1   17.489990   9.059998  25.270004   4.561001  -0.250000
2    4.779999   2.159996  -0.710007  -0.028000   8.679993
3    7.570007  15.130005  -9.800003  10.084999 -16.159996
4   16.720001   9.779999  29.919998   4.467999   5.070000
5    2.479996   3.319992  22.050018   4.427002  -0.010002
6   -8.580002   4.330002  -8.930023   2.625999   3.950005
7  -16.659988 -10.889992   3.900024  -5.855999   6.389999
8   -0.440002   5.969994 -24.820007  -2.719002 -11.730003
9   19.179993  13.000000  18.729980   5.990002  -3.110001
10   2.580002   5.850006  -5.859985   2.751999  -2.759995
11  -8.130005   3.259995 -11.680008  12.005001   2.829994
12  -3.649994  21.559998   7.680008  17.584999   1.709999
13  -9.270004   3.620010  40.769989  11.244003  11.720001
14  -1.149994  -5.380005 -27.959991  -3.954002   2.029999
15  21.919998   1.440002  -7.819992  23.231003  -1.009995
16  18.369995  16.809998  45.879990  13.906998  -2.139999
17  18.260010  -5.320007  10.589996  -5.550003   2.519997

In [22]:
# creating MX portfolio
mx_portfolio_df = pd.DataFrame()
mx_portfolio_df['GENIUS21.MX'] = gen21_df.dropna()
mx_portfolio_df['IVVPESOISHRS.MX'] = ivv_df.dropna()
mx_portfolio_df['WALMEX.MX'] = walmex_df.dropna()
# calc Returns
# https://numpy.org/doc/stable/reference/generated/numpy.log.html
#mx_portfolio_df["GENIUS21.MX_Returns"] = np.log(gen21_df.div(gen21_df.shift(1)))
mx_portfolio_df["GENIUS21.MX_Returns"] = gen21_df.pct_change()
#mx_portfolio_df["IVVPESOISHRS.MX_Returns"] = np.log(ivv_df.div(ivv_df.shift(1)))
mx_portfolio_df["IVVPESOISHRS.MX_Returns"] = ivv_df.pct_change()
#mx_portfolio_df["WALMEX.MX_Returns"] = np.log(walmex_df.div(walmex_df.shift(1)))
mx_portfolio_df["WALMEX.MX_Returns"] = walmex_df.pct_change()

mx_portfolio_df

GENIUS21.MX  IVVPESOISHRS.MX  WALMEX.MX  GENIUS21.MX_Returns  \
Date                                                                       
2023-01-02    37.230000              NaN  69.150002                  NaN   
2023-01-03    36.799999        70.349998  69.250000            -0.011550   
2023-01-04    37.450001        70.769997  70.489998             0.017663   
2023-01-05    36.860001        70.339996  70.410004            -0.015754   
2023-01-06    37.730000        72.000000  72.449997             0.023603   
...                 ...              ...        ...                  ...   
2024-07-11    86.040001       106.400002  65.410004            -0.031735   
2024-07-12    85.309998       106.820000  68.029999            -0.008484   
2024-07-15    86.059998       106.919998  67.199997             0.008791   
2024-07-16    84.860001       107.709999  65.820000            -0.013944   
2024-07-17    81.809998       106.290001  64.779999            -0.035942   

            IVVPESOISHRS.MX_Returns  WALMEX.MX_Returns  
Date                                                    
2023-01-02                      NaN                NaN  
2023-01-03                      NaN           0.001446  
2023-01-04                 0.005970           0.017906  
2023-01-05                -0.006076          -0.001135  
2023-01-06                 0.023600           0.028973  
...                             ...                ...  
2024-07-11                -0.004491           0.015841  
2024-07-12                 0.003947           0.040055  
2024-07-15                 0.000936          -0.012201  
2024-07-16                 0.007389          -0.020536  
2024-07-17                -0.013184          -0.015801  

[384 rows x 6 columns]

In [23]:
# create MX dayly df difference with values
mx_portfolio_diff_df = pd.DataFrame()
ticks_mx = ['GENIUS21.MX', 'IVVPESOISHRS.MX', 'WALMEX.MX']

for t in ticks_mx:
    # transform into differences dt
    j = 0
    tmp_l = []
    for i in mx_portfolio_df[t].index:
        #print(t, i, j)
        if j > 0:
            # look for consecutive cells
            #print(us_portfolio_df[t].iloc[j-1] - us_portfolio_df[t].iloc[j])
            tmp_l.append(mx_portfolio_df[t].iloc[j] - mx_portfolio_df[t].iloc[j-1])
        j = j + 1
    mx_portfolio_diff_df = pd.concat([mx_portfolio_diff_df, pd.DataFrame(tmp_l, columns=[t])], axis=1)

mx_portfolio_diff_df

GENIUS21.MX  IVVPESOISHRS.MX  WALMEX.MX
0      -0.430000              NaN   0.099998
1       0.650002         0.419998   1.239998
2      -0.590000        -0.430000  -0.079994
3       0.869999         1.660004   2.039993
4       0.180000         0.150002  -0.079994
..           ...              ...        ...
378    -2.820000        -0.479996   1.020004
379    -0.730003         0.419998   2.619995
380     0.750000         0.099998  -0.830002
381    -1.199997         0.790001  -1.379997
382    -3.050003        -1.419998  -1.040001

[383 rows x 3 columns]

In [24]:
# plot mx portfolio
cf.set_config_file(offline = True)
mx_portfolio_diff_df.iplot()

In [25]:
# plot mx portfolio
cf.set_config_file(offline = True)
mx_portfolio_df[ticks_mx].iplot()

In [26]:
# create MX monthly df difference with values
# = aapl_df.groupby(aapl_df.index.month).tail(1)
#   = amzn_df.groupby(amzn_df.index.month).tail(1)
#    = fdx_df.groupby(fdx_df.index.month).tail(1)
#   = nvda_df.groupby(nvda_df.index.month).tail(1)
#    = xom_df.groupby(xom_df.index.month).tail(1)
#  = gen21_df.groupby(gen21_df.index.month).tail(1)
#    = ivv_df.groupby(ivv_df.index.month).tail(1)
# = walmex_df.groupby(walmex_df.index.month).tail(1)
mx_portfolio_m_df = pd.DataFrame()
mx_portfolio_m_diff_df = pd.DataFrame()
ticks_mx = ['GENIUS21.MX', 'IVVPESOISHRS.MX', 'WALMEX.MX']

mx_portfolio_m_df['GENIUS21.MX'] = gen21_df_m.dropna()
mx_portfolio_m_df['IVVPESOISHRS.MX'] = ivv_df_m.dropna()
mx_portfolio_m_df['WALMEX.MX'] = walmex_df_m.dropna()

for t in ticks_mx:
    # transform into differences dt
    tmp_l = []
    j = 0
    for i in mx_portfolio_m_df[t].index:
        #print(t, i, j)
        if j > 0:
            # look for consecutive cells
            #print(us_portfolio_df[t].iloc[j-1] - us_portfolio_df[t].iloc[j])
            tmp_l.append(mx_portfolio_m_df[t].iloc[j] - mx_portfolio_m_df[t].iloc[j-1])
        j = j + 1
    mx_portfolio_m_diff_df = pd.concat([mx_portfolio_m_diff_df, pd.DataFrame(tmp_l, columns=[t])], axis=1)

#tmp_portfolio_diff_df = pd.DataFrame(tmp_l, columns=[t])
#us_portfolio_diff_df = pd.concat([us_portfolio_diff_df, tmp_portfolio_diff_df], ignore_index=True, sort=False)

mx_portfolio_m_diff_df

GENIUS21.MX  IVVPESOISHRS.MX  WALMEX.MX
0     -0.820000        -1.529999  -1.450005
1      2.939999         1.879997  -0.099998
2     -0.259998         1.290001   0.470001
3      3.770000         0.910004  -5.180000
4      2.389999         4.779999   0.599998
5      1.379997         1.750000   1.860001
6      0.880001        -1.279999  -2.529999
7     -2.969997        -4.080002  -1.440002
8      1.480000        -1.650002  -1.180000
9      4.189999         7.400002   3.779999
10     0.720001         4.340004   3.010002
11     6.539997         1.469994  -0.300003
12     8.520000         4.920006  -3.009995
13    -1.419998         0.059998  -4.230003
14     6.949997         2.839996  -0.099998
15    11.580002         5.480003  -1.259998
16    -4.830002         2.540001   2.289997

In [27]:
mx_portfolio_m_df

GENIUS21.MX  IVVPESOISHRS.MX  WALMEX.MX
Date                                               
2023-01-31    40.770000        75.169998  73.550003
2023-02-28    39.950001        73.639999  72.099998
2023-03-31    42.889999        75.519997  72.000000
2023-04-28    42.630001        76.809998  72.470001
2023-05-31    46.400002        77.720001  67.290001
2023-06-30    48.790001        82.500000  67.889999
2023-07-31    50.169998        84.250000  69.750000
2023-08-31    51.049999        82.970001  67.220001
2023-09-29    48.080002        78.889999  65.779999
2023-10-31    49.560001        77.239998  64.599998
2023-11-30    53.750000        84.639999  68.379997
2023-12-29    54.470001        88.980003  71.389999
2024-01-31    61.009998        90.449997  71.089996
2024-02-29    69.529999        95.370003  68.080002
2024-04-30    68.110001        95.430000  63.849998
2024-05-31    75.059998        98.269997  63.750000
2024-06-28    86.639999       103.750000  62.490002
2024-07-17    81.809998       106.290001  64.779999

In [72]:
# create returns in monthly dataset
mx_portfolio_m_df = mx_portfolio_m_df.drop('GENIUS21.MX_Returns', axis=1)
mx_portfolio_m_df = mx_portfolio_m_df.drop('IVVPESOISHRS.MX_Returns', axis=1)
mx_portfolio_m_df = mx_portfolio_m_df.drop('WALMEX.MX_Returns', axis=1)
mx_portfolio_m_df.dropna()
#dataset.resample('M')['wind'].sum()
#mx_portfolio_df.resample('M')['GENIUS21.MX_Returns'].sum()
#df.groupby(df.your_date_column.dt.month)['values_column'].sum()
#type(mx_portfolio_df.groupby(mx_portfolio_df.index.month)['GENIUS21.MX_Returns'].sum())
# create DF and thrn concat it to existent
tmp_m_l = (mx_portfolio_df.groupby(mx_portfolio_df.index.month)['GENIUS21.MX_Returns'].sum())
mx_portfolio_m_df['GENIUS21.MX_Returns'] = tmp_m_l
#mx_portfolio_m_df['GENIUS21.MX_Returns']     = mx_portfolio_df.groupby(mx_portfolio_df.index.month)['GENIUS21.MX_Returns'].sum()
#mx_portfolio_m_df['IVVPESOISHRS.MX_Returns'] = mx_portfolio_df.groupby(mx_portfolio_df.index.month)['IVVPESOISHRS.MX_Returns'].sum()
#mx_portfolio_m_df['WALMEX.MX_Returns']       = mx_portfolio_df.groupby(mx_portfolio_df.index.month)['WALMEX.MX_Returns'].sum()

mx_portfolio_m_df

GENIUS21.MX  IVVPESOISHRS.MX  WALMEX.MX  GENIUS21.MX_Returns
0     51.049999        82.970001  67.220001                  NaN
1     48.080002        78.889999  65.779999             0.208046
2     49.560001        77.239998  64.599998             0.117285
3     53.750000        84.639999  68.379997             0.089298
4     54.470001        88.980003  71.389999            -0.037503
5     61.009998        90.449997  71.089996             0.185656
6     69.529999        95.370003  68.080002             0.200311
7     68.110001        95.430000  63.849998             0.049529
8     75.059998        98.269997  63.750000             0.021057
9     86.639999       103.750000  62.490002            -0.058481
10    88.459999       106.139999  64.059998             0.033007
11          NaN              NaN        NaN             0.081862
12          NaN              NaN        NaN             0.014468
13          NaN              NaN        NaN                  NaN
14          NaN              NaN        NaN                  NaN
15          NaN              NaN        NaN                  NaN
16          NaN              NaN        NaN                  NaN
17          NaN              NaN        NaN                  NaN
18          NaN              NaN        NaN                  NaN
19          NaN              NaN        NaN                  NaN
20          NaN              NaN        NaN                  NaN
21          NaN              NaN        NaN                  NaN
22          NaN              NaN        NaN                  NaN
23          NaN              NaN        NaN                  NaN
24          NaN              NaN        NaN                  NaN
25          NaN              NaN        NaN                  NaN
26          NaN              NaN        NaN                  NaN
27          NaN              NaN        NaN                  NaN
28          NaN              NaN        NaN                  NaN
29          NaN              NaN        NaN                  NaN
30          NaN              NaN        NaN                  NaN
31          NaN              NaN        NaN                  NaN
32          NaN              NaN        NaN                  NaN
33          NaN              NaN        NaN                  NaN
34          NaN              NaN        NaN                  NaN
35          NaN              NaN        NaN                  NaN
36          NaN              NaN        NaN                  NaN
37          NaN              NaN        NaN                  NaN
38          NaN              NaN        NaN                  NaN
39          NaN              NaN        NaN                  NaN
40          NaN              NaN        NaN                  NaN
41          NaN              NaN        NaN                  NaN
42          NaN              NaN        NaN                  NaN
43          NaN              NaN        NaN                  NaN
44          NaN              NaN        NaN                  NaN
45          NaN              NaN        NaN                  NaN
46          NaN              NaN        NaN                  NaN
47          NaN              NaN        NaN                  NaN
48          NaN              NaN        NaN                  NaN
49          NaN              NaN        NaN                  NaN
50          NaN              NaN        NaN                  NaN
51          NaN              NaN        NaN                  NaN
52          NaN              NaN        NaN                  NaN
53          NaN              NaN        NaN                  NaN
54          NaN              NaN        NaN                  NaN
55          NaN              NaN        NaN                  NaN
56          NaN              NaN        NaN                  NaN
57          NaN              NaN        NaN                  NaN
58          NaN              NaN        NaN                  NaN

In [28]:
# create us analysis portfolio dayly info
index_l = ['Day_outcome', 'Month_outcome', 'Year_outcome', 'Year_var', 'Cov_year_mkt']
us_portfolio_analysis_df = pd.DataFrame(index = index_l)
#print(us_portfolio_df['AAPL'].mean())
us_portfolio_analysis_df['AAPL'] = us_portfolio_diff_df['AAPL'].mean()
us_portfolio_analysis_df['AMZN'] = us_portfolio_diff_df['AMZN'].mean()
us_portfolio_analysis_df['FDX'] = us_portfolio_diff_df['FDX'].mean()
us_portfolio_analysis_df['NVDA'] = us_portfolio_diff_df['NVDA'].mean()
us_portfolio_analysis_df['XOM']  = us_portfolio_diff_df['XOM'].mean()
us_portfolio_analysis_df


AAPL      AMZN      FDX      NVDA       XOM
Day_outcome    0.269636  0.265221  0.34587  0.269286  0.028909
Month_outcome  0.269636  0.265221  0.34587  0.269286  0.028909
Year_outcome   0.269636  0.265221  0.34587  0.269286  0.028909
Year_var       0.269636  0.265221  0.34587  0.269286  0.028909
Cov_year_mkt   0.269636  0.265221  0.34587  0.269286  0.028909

In [29]:
# us portfolio monthly info
#aapl_df_m = aapl_df.groupby(aapl_df.index.month).tail(1)
#print(aapl_df_m.mean())
us_portfolio_analysis_df['AAPL']['Month_outcome'] = us_portfolio_m_diff_df['AAPL'].mean()
us_portfolio_analysis_df['AAPL']['Year_outcome']  = us_portfolio_m_diff_df['AAPL'].mean()*12
us_portfolio_analysis_df['AAPL']['Year_var']      = us_portfolio_df["AAPL_Returns"].var()
#us_portfolio_analysis_df['AAPL']['Cov_year_mkt']  = us_portfolio_df["AAPL_Returns"].cov()
us_portfolio_analysis_df['AAPL']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["AAPL_Returns"].var())
#amzn_df_m   = amzn_df.groupby(amzn_df.index.month).tail(1)
us_portfolio_analysis_df['AMZN']['Month_outcome'] = us_portfolio_m_diff_df['AMZN'].mean()
us_portfolio_analysis_df['AMZN']['Year_outcome']  = us_portfolio_m_diff_df['AMZN'].mean()*12
us_portfolio_analysis_df['AMZN']['Year_var']      = us_portfolio_df["AMZN_Returns"].var()
#us_portfolio_analysis_df['AMZN']['Cov_year_mkt']  = us_portfolio_df["AMZN_Returns"].cov()
us_portfolio_analysis_df['AMZN']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["AMZN_Returns"].var())
#fdx_df_m   = fdx_df.groupby(amzn_df.index.month).tail(1)
us_portfolio_analysis_df['FDX']['Month_outcome'] = us_portfolio_m_diff_df['FDX'].mean()
us_portfolio_analysis_df['FDX']['Year_outcome']  = us_portfolio_m_diff_df['FDX'].mean()*12
us_portfolio_analysis_df['FDX']['Year_var']      = us_portfolio_df["FDX_Returns"].var()
#us_portfolio_analysis_df['FDX']['Cov_year_mkt']  = us_portfolio_df["FDX_Returns"].cov()
us_portfolio_analysis_df['FDX']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["FDX_Returns"].var())
#nvda_df_m   = nvda_df.groupby(nvda_df.index.month).tail(1)
us_portfolio_analysis_df['NVDA']['Month_outcome'] = us_portfolio_m_diff_df['NVDA'].mean()
us_portfolio_analysis_df['NVDA']['Year_outcome']  = us_portfolio_m_diff_df['NVDA'].mean()*12
us_portfolio_analysis_df['NVDA']['Year_var']      = us_portfolio_df["NVDA_Returns"].var()
#us_portfolio_analysis_df['NVDA']['Cov_year_mkt']  = us_portfolio_df["NVDA_Returns"].cov()
us_portfolio_analysis_df['NVDA']['Cov_year_mkt']  = np.sqrt(us_portfolio_df["NVDA_Returns"].var())
#xom_df_m    = xom_df.groupby(xom_df.index.month).tail(1)
us_portfolio_analysis_df['XOM']['Month_outcome']  = us_portfolio_m_diff_df['XOM'].mean()
us_portfolio_analysis_df['XOM']['Year_outcome']   = us_portfolio_m_diff_df['XOM'].mean()*12
us_portfolio_analysis_df['XOM']['Year_var']       = us_portfolio_df["XOM_Returns"].var()
#us_portfolio_analysis_df['XOM']['Cov_year_mkt']   = us_portfolio_df["XOM_Returns"].cov()
us_portfolio_analysis_df['XOM']['Cov_year_mkt']   = np.sqrt(us_portfolio_df["XOM_Returns"].var())

us_portfolio_analysis_df

AAPL       AMZN        FDX       NVDA       XOM
Day_outcome     0.269636   0.265221   0.345870   0.269286  0.028909
Month_outcome   4.699445   4.711111   6.476111   5.469611  0.090555
Year_outcome   56.393341  56.533330  77.713328  65.635331  1.086665
Year_var        0.000189   0.000366   0.000318   0.000986  0.000207
Cov_year_mkt    0.013733   0.019122   0.017825   0.031404  0.014385

the analysis explains that I can expect $0.27 dayly from AAPL, $4.70 monthly or $56.39 yearly for any dollar we are investing.
#https://www.investopedia.com/terms/v/variance.asp#:~:text=The%20term%20variance%20refers%20to,by%20this%20symbol%3A%20%CF%832.
The term variance refers to a statistical measurement of the spread between numbers in a data set. More specifically, variance measures how far each number in the set is from the mean (average), and thus from every other number in the set. Variance is often depicted by this symbol: σ2. It is used by both analysts and traders to determine volatility and market security.
in the year AAPL shows 0.0189\% of volatility with 1.373\% variation for this stock.
NVDA is more risky with 3.14\% of variation in the price.

In [27]:
# mx portfolio monthly info
index_l = ['Day_outcome', 'Month_outcome', 'Year_outcome', 'Year_var', 'Cov_year_mkt']
mx_portfolio_analysis_df = pd.DataFrame(index = index_l)

mx_portfolio_analysis_df['GENIUS21.MX'] = mx_portfolio_diff_df['GENIUS21.MX'].mean()
mx_portfolio_analysis_df['IVVPESOISHRS.MX'] = mx_portfolio_diff_df['IVVPESOISHRS.MX'].mean()
mx_portfolio_analysis_df['WALMEX.MX'] = mx_portfolio_diff_df['WALMEX.MX'].mean()
# GENIUS21.MX
mx_portfolio_analysis_df['GENIUS21.MX']['Month_outcome'] = mx_portfolio_m_diff_df['GENIUS21.MX'].mean()
mx_portfolio_analysis_df['GENIUS21.MX']['Year_outcome']  = mx_portfolio_m_diff_df['GENIUS21.MX'].mean()*12
mx_portfolio_analysis_df['GENIUS21.MX']['Year_var']      = mx_portfolio_df["GENIUS21.MX_Returns"].var()
#mx_portfolio_analysis_df['GENIUS21.MX']['Cov_year_mkt'] = mx_portfolio_df["GENIUS21.MX_Returns"].cov()
mx_portfolio_analysis_df['GENIUS21.MX']['Cov_year_mkt']  = np.sqrt(mx_portfolio_df["GENIUS21.MX_Returns"].var())
# IVVPESOISHRS.MX
mx_portfolio_analysis_df['IVVPESOISHRS.MX']['Month_outcome'] = mx_portfolio_m_diff_df['IVVPESOISHRS.MX'].mean()
mx_portfolio_analysis_df['IVVPESOISHRS.MX']['Year_outcome']  = mx_portfolio_m_diff_df['IVVPESOISHRS.MX'].mean()*12
mx_portfolio_analysis_df['IVVPESOISHRS.MX']['Year_var']      = mx_portfolio_df["IVVPESOISHRS.MX_Returns"].var()
#mx_portfolio_analysis_df['IVVPESOISHRS.MX']['Cov_year_mkt'] = mx_portfolio_df["IVVPESOISHRS.MX_Returns"].cov()
mx_portfolio_analysis_df['IVVPESOISHRS.MX']['Cov_year_mkt']  = np.sqrt(mx_portfolio_df["IVVPESOISHRS.MX_Returns"].var())
# WALMEX.MX
mx_portfolio_analysis_df['WALMEX.MX']['Month_outcome'] = mx_portfolio_m_diff_df['WALMEX.MX'].mean()
mx_portfolio_analysis_df['WALMEX.MX']['Year_outcome']  = mx_portfolio_m_diff_df['WALMEX.MX'].mean()*12
mx_portfolio_analysis_df['WALMEX.MX']['Year_var']      = mx_portfolio_df["WALMEX.MX_Returns"].var()
#mx_portfolio_analysis_df['WALMEX.MX']['Cov_year_mkt']  = mx_portfolio_df["WALMEX.MX_Returns"].cov()
mx_portfolio_analysis_df['WALMEX.MX']['Cov_year_mkt']  = np.sqrt(mx_portfolio_df["WALMEX.MX_Returns"].var())

mx_portfolio_analysis_df

GENIUS21.MX  IVVPESOISHRS.MX  WALMEX.MX
Day_outcome       0.135889         0.105589  -0.013501
Month_outcome     3.741000         2.317000  -0.316000
Year_outcome     44.892000        27.803998  -3.792004
Year_var          0.000204         0.000071   0.000216
Cov_year_mkt      0.014298         0.008447   0.014695